### PR TITLE
Unified picture branch

### PR DESCRIPTION
**Pull request by Byvire**
Incorporates Jon's and my SVG-drawing submodules

### CODE REVIEW

**Comment by Byvire** [comment]
@Agizin I'd like to make our goal be to merge this branch into master instead of either of our individual PRs

File: algviz/picture/recursive/rec_layout.py [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+
+    def __init__(self, svg_hint=None, choose_child_cls=None, ptr_targets=None):
+        self.svg_hint = svg_hint
+        self._choose_child_cls = choose_child_cls
+        self._placements = {}  # {placement: coordinate}
+        self._children = {}  # Like placements, but for sub-pictures instead of primitives
+        self._decorations = []  # these have no coordinates
+        # ptr_targets is a map from DataStructure instances to the
+        # PictureElement instances that have been placed to represent them.
+        self._ptr_targets = {} if ptr_targets is None else ptr_targets
```
Comment (Agizin) [inline algviz/picture/recursive/rec_layout.py]: ptr_targets should be renamed as it doesn't have anything to do with pointer structures (intuitively, at least).

File: algviz/picture/recursive/layout_dispatch.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+from . import rec_layout
+from algviz.parser import structures
+
+def map_metadata_key_to_layout_subclass(config):
```
[the remark this answers is not in the document]
Comment (Byvire) [reply algviz/picture/recursive/layout_dispatch.py]: Yeah, this file is sort of a sham right now.  It just ignores all the config.

File: algviz/picture/recursive/svg.py [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+class SVGHint(
+        StringHint,
+        MarginHint,
+        PointerHint,
+        NullHint,
+        ArrayFrameHint,
+):
+    """In order to create a Layout, it's useful to know what size the components of the picture should be."""
+    
+class DelegatingSVGHint(SVGHint):
```
[the remark this answers is not in the document]
Comment (Byvire) [reply algviz/picture/recursive/svg.py]: Just separating implementation from interface.  Currently there's only one type of Hint, though.

File: algviz/picture/recursive/svg.py [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+            anchors.from_top_left_corner(null, coord, anch)
+            for anch in [anchors.Anchor.top,
+                         anchors.Anchor.right,
+                         anchors.Anchor.bottom,
+                         anchors.Anchor.left])))
+
+class NodeElementDrawer(RectangularElementDrawer):
+    pass
+
+class EllipseNodeElementDrawer(RectangularElementDrawer):
```
[the remark this answers is not in the document]
Comment (Byvire) [reply algviz/picture/recursive/svg.py]: Thanks

File: algviz/picture/recursive/svg.py [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+    @property
+    @abc.abstractmethod
+    def array_cell_sep(self):
+        """Amount of space to leave between adjacent cells of an array"""
+
+class PointerHint(metaclass=abc.ABCMeta):
+    @property
+    @abc.abstractmethod
+    def pointer_size(self):
+        """The (width, height) of a Pointer"""
```
[the remark this answers is not in the document]
Comment (Byvire) [reply algviz/picture/recursive/svg.py]: Yeah, it's the little dot or whatever that the arrow comes from.  I've seen dots used for this a lot, so I made this one a dot.

File: algviz/picture/recursive/rec_layout.py (hunk 1 above)
Comment (Byvire) [reply algviz/picture/recursive/rec_layout.py]: Yeah, it's a mapping from DataStructure instances to the corresponding Element instances.  I named it after how I use it, not after what it is.  But I suppose that's bad.

File: algviz/picture/recursive/svg.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+        svg_doc.add(svg_doc.line(src, dst, style="stroke:rgb(50,50,50);stroke-width:2"))
+        self.draw_arrowhead(dst, self.angle(src, dst), svg_doc)
+
+class SplineArrowElementDrawer(StraightArrowElementDrawer):
+    pass  # TODO -- splines instead
+
+class IncompleteLayoutError(Exception):
+    pass
+
+class FullDrawer:
```
[the remark this answers is not in the document]
Comment (Byvire) [reply algviz/picture/recursive/svg.py]: ... I should just change it to Painter

File: algviz/picture/recursive/svg.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Agizin) [reply algviz/picture/recursive/svg.py]: Ah, you might want to note that. In general, when you refer to drawing a pointer in your code, is it just the pointer dot (and then the arrow is drawn separately)?

**Comment by Byvire** [comment]
@Agizin FYI, `fabulous` still chokes when a tree has `Null` children (placeholders).  You can check with:
```bash
cat algviz/picture/test_cases/binary_tree.json | algviz_rec_draw - smallest_random_bin_tree.svg --var target -m fabulous
```

**Comment by Agizin** [comment]
@Byvire Do you get an error when drawing a tree with null children? It's worked fine when I've tested.

Note that I'm purposefully ignoring null children currently (i.e. not drawing a placeholder at that location). Rigid n-ary trees are purposefully not drawn by default. I'll add either a setting or a subclass of TreePicture that supports these drawings later (but not before demo day).

**Comment by Agizin** [comment]
@Byvire I see what the problem is. Why are your representing placeholders using `NullType` rather than `None`?

**Pull request merged**